src/Type/UnionType.php

```diff
@@ -92,15 +93,36 @@ public function getTypes(): array
 	public function filterTypes(callable $filterCb): Type
 	{
 		$newTypes = [];
+		$changed = false;
 		foreach ($this->getTypes() as $innerType) {
 			if (!$filterCb($innerType)) {
+				$changed = true;
 				continue;
 			}
 
 			$newTypes[] = $innerType;
 		}
 
-		return TypeCombinator::union(...$newTypes);
+		if (!$changed) {
+			return $this;
+		}
+
+		$result = TypeCombinator::union(...$newTypes);
+		if ($this instanceof BenevolentUnionType) {
+			$result = TypeUtils::toBenevolentUnion($result);
+		}
+		if ($this instanceof TemplateType) {
```

Comment:
I'd rather have this logic in the subtypes in overridden methods.

Reply:
Done